### PR TITLE
Object Storage: support Baidu Object Storage (BOS) as provider

### PR DESCRIPTION
**Pull request by tianyuansun**
## Changes

Baidu Cloud is a public cloud provider in China. [Baidu Cloud](https://cloud.baidu.com/).
Its object storage product is BOS (Baidu Object Storage).

We are testing Thanos with BOS as storage backend.
This pr implements a BOS client and it seems to be ok.

I would love to become the maintainer for the BOS provider and keep development for Thanos.

## Verification

here is acceptance_e2e_test result
image

go test pkg/objstore/objtesting/acceptance_e2e_test.go pkg/objstore/objtesting/foreach.go
ok  	command-line-arguments	3.275s

### CODE REVIEW

**Comment by jojohappy** [comment]
@tianyuansun Thank you for your PR! Awesome! Will review today later.

File: pkg/objstore/bos/bos.go [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+				Delimiter:  dirDelim,
+				Marker:     marker,
+				MaxKeys:    1000,
+			}, nil)
+			if err != nil {
+				select {
+				case objectsCh <- objectInfo{
+					err: err,
+				}:
+				case <-ctx.Done():
```
Comment (jojohappy) [inline pkg/objstore/bos/bos.go]: Maybe we don't need select here, we can check `ctx.Err()` to determine the iterator has been canceled.
```
for {
  if ctx.Err() != nil {
    return
  }
}
```

Further information is [here](https://golang.org/pkg/context/#Context).

File: pkg/objstore/bos/bos.go [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+				}
+				return
+			}
+
+			for _, object := range result.Contents {
+				select {
+				case objectsCh <- objectInfo{
+					key: object.Key,
+				}:
+				case <-ctx.Done():
```
Comment (jojohappy) [inline pkg/objstore/bos/bos.go]: ditto.

File: pkg/objstore/bos/bos.go [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+			}
+
+			// The result of CommonPrefixes contains the objects
+			// that have the same keys between Prefix and the key specified by delimiter.
+			for _, obj := range result.GetCommonPrefixes() {
+				select {
+				case objectsCh <- objectInfo{
+					key: obj,
+				}:
+				case <-ctx.Done():
```
Comment (jojohappy) [inline pkg/objstore/bos/bos.go]: ditto.

File: pkg/objstore/bos/bos.go [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+// GetRange returns a new range reader for the given object name and range.
+func (b *Bucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	return b.getRange(ctx, b.name, name, off, length)
+}
+
+// Exists checks if the given object exists in the bucket.
+// TODO(bplotka): Consider removing Exists in favor of helper that do Get & IsObjNotFoundErr (less code to maintain).
+func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
+	obj, err := b.getRange(ctx, b.name, name, 0, 2)
```
Comment (jojohappy) [inline pkg/objstore/bos/bos.go]: Just curious, why did you use `getRange` to check the object? Could we use the `GetObjectMeta` API?

File: pkg/objstore/bos/bos.go [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+func (b *Bucket) Delete(ctx context.Context, name string) error {
+	if err := b.client.DeleteObject(b.name, name, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	if _, err := b.client.PutObject(b.name, name, r, nil, nil); err != nil {
```
Comment (jojohappy) [inline pkg/objstore/bos/bos.go]: I saw the official documents that the sample upload interface could not upload the file large than 5GB, sometimes the tsdb files (index) would exceed 5GB. Could you double check this interface would not be impact by the limitation?

File: pkg/objstore/bos/bos.go [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+
+	"github.com/baidu/baiducloud-sdk-go/bce"
+	"github.com/baidu/baiducloud-sdk-go/bos"
+	"github.com/go-kit/kit/log"
+	"github.com/improbable-eng/thanos/pkg/objstore"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+)
+
+// DirDelim is the delimiter used to model a directory structure in an object store bucket.
```
Comment (jojohappy) [inline pkg/objstore/bos/bos.go]: Remove this line, the parameter is not exported.

File: pkg/objstore/bos/bos.go [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	if _, err := b.client.PutObject(b.name, name, r, nil, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Iter calls f for each entry in the given directory (not recursive.). The argument to f is the full
```
Comment (jojohappy) [inline pkg/objstore/bos/bos.go]:
```suggestion
// Iter calls f for each entry in the given directory (not recursive). The argument to f is the full
```

File: pkg/objstore/bos/bos.go [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+	"net/http"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/baidu/baiducloud-sdk-go/bce"
+	"github.com/baidu/baiducloud-sdk-go/bos"
+	"github.com/go-kit/kit/log"
+	"github.com/improbable-eng/thanos/pkg/objstore"
```
Comment (jojohappy) [inline pkg/objstore/bos/bos.go]: Move this line here:
```
"time"

"github.com/improbable-eng/thanos/pkg/objstore"

"github.com/baidu/baiducloud-sdk-go/bce"
```

**Comment by bwplotka** [comment]
Any plans to update this PR? @tianyuansun

**Pull request closed without merging**